### ISSUE TITLE
Determine JSON model for Index Item

### ISSUE THREAD

**Issue by daggerhart** [copperdogma/dnd-phb-5e-index#10]
@copperdogma @Falanyx

The current item model in my PR #9 

```
{
    "name": "", // string
    "pages": [ "" ], // array of integers or strings
    "children": [ {} ] // array of items
}
```

@copperdogma mentioned "canonical pages numbers", and I also thought maybe we wanted to break the "see also"s into separate data as well.  Maybe we add a "description" property, and tuck "see also"s into that?

**Comment by copperdogma** [comment]
Well, I can think of a few ways to do canonical references. I THINK (but I'm not sure) there may be multiple canonical numbers for a single entry.
1. Have all page numbers be string, not numbers, and just wrap them in markdown **. Like "**16**".
2. Have a second entry for "canonical-pages" and just list out the canonical ones again. The output would need to merge the two and highlight the canonical ones.
3. Simplify even more. I'm not sure we need pages split out into an array. It's not really data because we don't do anything with it except concat it back together and output it, so why not just leave it as a whole string instead of splitting it into an array in the first place. And if we do that we can definitely use the Markdown syntax to highlight canonical.

Thoughts?

**Comment by daggerhart** [comment]
I like option 2 here.  I'd rather have data that I can convert easily into content if i need it, than content I have convert to data.  Err- on the side of 'data'.

What did you think about the "description" property?  Or another way to represent the "See also..." data?

Both of those together could give us: 

```
{
    "name": "", // string
    "description": "", // (optional) string
    "pages": [ "" ], // array of integers or strings
    "canonical": [ "" ], // (optional) array of integers or strings that indicate most-important
    "children": [ {} ] // array of items
}
```

Example:

```
{
    "name": "class",
    "description": "See also specific class entries"
    "pages": [ 11, 45 ],
    "canonical": [ 45 ]
    "children": [
        { "name": "choosing a", "pages": [11] },
        { "name": "features", "pages": [ 11, 15 ] },
        { "name": "proficiencies", "pages": [12] },
        { "name": "quick build", "pages": [11] }
    ]
}
```

**Comment by ndguarino** [comment]
The current proposed data structure seems fine, though I will note as I did somewhere in #8 (I think. I could be wrong) that it would be, while slightly less readable in json format, easier to search/handle the data if the children were references rather than data structures in their own right. It's (marginally) easier to extrapolate and index relationships from a flat structure than to recurse into a heirarchical structure.

Either should be fine, though.

**Comment by daggerhart** [comment]
The more I think about it the more I think @Falanyx is right here. 

Something like?

```
{
    "key": "ability-check",
    "name": "ability check",
    "parent": null,
    "pages": [ 7, 12, 173-179, 186 ],
    "canonical": [ 173 ]
},
    { "parent": "ability-check", "name": "contest", "pages": [ 174  ] },
    { "parent": "ability-check", "name": "group", "pages": [ 175  ] },
    { "parent": "ability-check", "name": "passive", "pages": [ 175  ] },
    { "parent": "ability-check", "name": "skill", "pages": [ 174-175 ] },
    { "parent": "ability-check", "name": "skills with different abilities (variant)", "pages": [ 175  ] },
    { "parent": "ability-check", "name": "working together", "pages": [ 175  ] },
```

That's readable enough for me.

**Comment by ndguarino** [comment]
Yeah, that's about what I was thinking. It's actually pretty close to what I had originally, I think, albeit with the pages/canonical separated in yours (as they should be).

**Comment by copperdogma** [comment]
Yup I like it! I just wanted to nail down a format where we COULD add canonical. I'll do the actual work for that section later.

For now.. I think we're ready to convert from .txt=master to .json=master.

I've revised the .txt file a few times since, so we need to do the following:
1. Regenerate a new json file based off the current .txt version.
2. Check it for accuracy and fix by hand where necessary until it's a perfect master.
3. Start using it! Keep building the web version using the new .json master and create a tool to generate the .txt version from the .json.

Note for item 3: I dislike command-line build tools. They require taking into consideration the operating system and/or cross-platform build tools, the commands are arcane, and debugging is a bit. I much prefer to just create an ugly website with an "import json" button and a "generate txt" button with a big text box for the output. Then you copy and paste it into your .txt file. Easy, cross-platform, can be run off your machine or online, and trivial to debug using a browser.

I'm interested in building that json -> txt web page. Anyone else want to volunteer for any of the other tasks? If not let me know and I'll do them too!

Thoughts?

**Comment by copperdogma** [comment]
Okay I've created a pull request for the json->txt converter. Take a look and let me know what you think. Very basic, but very straightforward and seems to work just fine.

**Comment by copperdogma** [comment]
Another point I think we should consider: Adding a version number to the json file would be a great thing to be able to output in final formats. Then people can check and see if they're using the latest version or not.
